src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs

```diff
@@ -0,0 +1,1867 @@
+        /// - ThrowOnExisting: If the key already exists, throws an exception, e.g. Add(key, value)
+        /// </param>
+        /// <returns>true if the collection was updated; otherwise, false.</returns>
+        private bool TryInsert(int index, TKey key, TValue value, InsertionBehavior behavior)
+        {
+            // Search for the key in the dictionary.
+            uint hashCode = 0, collisionCount = 0;
+            int i = IndexOf(key, ref hashCode, ref collisionCount);
+
+            // Handle the case where the key already exists, based on the requested behavior.
+            if (i >= 0)
+            {
+                Debug.Assert(_entries is not null);
+
+                switch (behavior)
+                {
+                    case InsertionBehavior.OverwriteExisting:
+                        Debug.Assert(index < 0, "Expected index to be unspecied when overwriting an existing key.");
+                        _entries[i].Value = value;
+                        return true;
+
+                    case InsertionBehavior.ThrowOnExisting:
+                        ThrowHelper.ThrowDuplicateKey(key);
+                        break;
+
+                    default:
+                        Debug.Assert(behavior is InsertionBehavior.IgnoreInsertion, $"Unknown behavior: {behavior}");
+                        Debug.Assert(index < 0, "Expected index to be unspecied when ignoring a duplicate key.");
+                        return false;
+                }
+            }
+
+            // The key doesn't exist. If a non-negative index was provided, that is the desired index at which to insert,
+            // which should have already been validated by the caller. If negative, we're appending.
+            if (index < 0)
+            {
+                index = _count;
+            }
+            Debug.Assert(index <= _count);
+
+            // Ensure the collection has been initialized.
+            if (_buckets is null)
+            {
+                EnsureBucketsAndEntriesInitialized(0);
+            }
+
+            // As we just initialized the collection, _entries must be non-null.
+            Entry[]? entries = _entries;
+            Debug.Assert(entries is not null);
+
+            // Grow capacity if necessary to accomodate the extra entry.
+            if (entries.Length == _count)
+            {
+                Resize(HashHelpers.ExpandPrime(entries.Length));
+                entries = _entries;
+            }
+
+            // The _entries array is ordered, so we need to insert the new entry at the specified index. That means
+            // not only shifting up all elements at that index and higher, but also updating the buckets and chains
+            // to record the newly updated indices.
+            for (i = _count - 1; i >= index; --i)
+            {
+                entries[i + 1] = entries[i];
```

Comment:
Is there a chance we could perf improvement if the shifting of the entries array used block copying?

Reply:
Possibly. But we still need to walk the entries individually to fix up the chains. We can more easily experiment with it once it's in a daily.